File: solutions/1027. Longest Arithmetic Subsequence/1027.cpp
// Time complexity: O(n^2)
// Space complexity: O(n)
class Solution {
 public:
  int longestArithSeqLength(const vector<int>& nums) {
    const int n = nums.size();
    int ans = 0;
    vector<vector<int>> dp(n, vector<int>(1001));

    for (int i = 0; i < n; ++i)
      for (int j = 0; j < i; ++j) {
        const int k = nums[i] - nums[j] + 500;
        dp[i][k] = max(2, dp[j][k] + 1);
        ans = max(ans, dp[i][k]);
      }

    return ans;
  }
};

File: solutions/1027. Longest Arithmetic Subsequence/1027.rs
// Time complexity: O(n^2)
// Space complexity: O(n)
impl Solution {
    pub fn longest_artith_seq_length(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut ans = 0;
        let mut dp = vec![vec![0; 1001]; n];

        for i in 0..n {
            for j in 0..i {
                let k = (nums[i] - nums[j] + 500) as usize;
                dp[i][k] = 2.max(dp[j][k] + 1);
                ans = ans.max(dp[i][k]);
            }
        }

        ans
    }
}
